### PR TITLE
Naji/force http2 kubernetes fix

### PR DESCRIPTION
**Pull request by NajiObeid**
https://github.com/gravitational/teleport/pull/9294#issuecomment-1020808550

### CODE REVIEW

**Comment by russjones** [comment]
@NajiObeid Can you add test coverage to this PR?

File: lib/kube/proxy/forwarder.go
Comment (r0mant) [inline lib/kube/proxy/forwarder.go:1569]: Nit: Explain with a comment why we sometimes want to upgrade to HTTP2 and sometimes we don't.

**Review by r0mant — APPROVED** [review]
lgtm if you've made sure this works and the feedback is addressed (and please wait for a 2nd review as well).

File: lib/kube/proxy/forwarder.go
Comment (jimbishopp) [inline lib/kube/proxy/forwarder.go:1561]: nit: I like to see short godoc comments on all functions. It would help clarify their purpose and help new developers learn the codebase.

File: lib/kube/proxy/forwarder.go
Comment (jimbishopp) [reply lib/kube/proxy/forwarder.go:1058]: +1

**Review by jimbishopp — COMMENTED** [review]
Looks good pending test coverage.

**Review by r0mant — APPROVED** [review]

**Review by jimbishopp — APPROVED** [review]

**Comment by russjones** [comment]
`lib/kube` unit tests are consistently failing.

```
FAIL: github.com/gravitational/teleport/lib/kube/proxy
FAIL: github.com/gravitational/teleport/lib/kube/proxy.TestNewClusterSessionDirect
FAIL: github.com/gravitational/teleport/lib/kube/proxy.TestNewClusterSessionLocal
FAIL: github.com/gravitational/teleport/lib/kube/proxy.TestNewClusterSessionRemote
===================================================
OUTPUT github.com/gravitational/teleport/lib/kube/proxy.TestNewClusterSessionDirect
===================================================
=== RUN   TestNewClusterSessionDirect
time="2022-01-28T22:45:34Z" level=info msg="Generating TLS certificate {0xc0005a2f00 0xc000156540 CN=bob,OU=usage a+OU=usage b,O=group a+O=group b,POSTALCODE={\\\"trait a\\\":[\\\"b\\\"\\,\\\"c\\\"]},STREET=,L=principal a+L=principal b,ST=k8s group a+ST=k8s group b,1.3.9999.1.2=#130b6b38732067726f75702062,1.3.9999.1.2=#130b6b38732067726f75702061 2022-01-28 22:46:34.643553914 +0000 UTC m=+60.240996224 [] [] 5 []}." common_name=bob dns_names="[]" locality="[principal a principal b]" not_after="2022-01-28 22:46:34.643553914 +0000 UTC m=+60.240996224" org="[group a group b]" org_unit="[usage a usage b]" trace.component=ca
    forwarder_test.go:753: 
        	Error Trace:	forwarder_test.go:753
        	Error:      	Expected value not to be nil.
        	Test:       	TestNewClusterSessionDirect
--- FAIL: TestNewClusterSessionDirect (0.03s)
===================================================
OUTPUT github.com/gravitational/teleport/lib/kube/proxy.TestNewClusterSessionLocal
===================================================
=== RUN   TestNewClusterSessionLocal
    forwarder_test.go:687: 
        	Error Trace:	forwarder_test.go:687
        	Error:      	Expected value not to be nil.
        	Test:       	TestNewClusterSessionLocal
--- FAIL: TestNewClusterSessionLocal (0.00s)
===================================================
OUTPUT github.com/gravitational/teleport/lib/kube/proxy.TestNewClusterSessionRemote
===================================================
=== RUN   TestNewClusterSessionRemote
time="2022-01-28T22:45:34Z" level=info msg="Generating TLS certificate {0xc000591260 0xc0000ddaa0 CN=bob,OU=usage a+OU=usage b,O=group a+O=group b,POSTALCODE={\\\"trait a\\\":[\\\"b\\\"\\,\\\"c\\\"]},STREET=,L=principal a+L=principal b,ST=k8s group a+ST=k8s group b,1.3.9999.1.2=#130b6b38732067726f75702062,1.3.9999.1.2=#130b6b38732067726f75702061 2022-01-28 22:46:34.615029901 +0000 UTC m=+60.212472215 [] [] 5 []}." common_name=bob dns_names="[]" locality="[principal a principal b]" not_after="2022-01-28 22:46:34.615029901 +0000 UTC m=+60.212472215" org="[group a group b]" org_unit="[usage a usage b]" trace.component=ca
    forwarder_test.go:705: 
        	Error Trace:	forwarder_test.go:705
        	Error:      	Expected value not to be nil.
        	Test:       	TestNewClusterSessionRemote
--- FAIL: TestNewClusterSessionRemote (0.02s)
===================================================
```

**Comment by tcsc** [comment]
The `TestKube/Deny` and `TestKube/Disconnect/#01_iter_0` integration tests are persistently failing for me when executed against a GKE cluster:

```
pass: github.com/gravitational/teleport/integration.TestKube/Exec
fail: github.com/gravitational/teleport/integration.TestKube/Deny
pass: github.com/gravitational/teleport/integration.TestKube/PortForward
pass: github.com/gravitational/teleport/integration.TestKube/TransportProtocol
pass: github.com/gravitational/teleport/integration.TestKube/TrustedClustersClientCert
pass: github.com/gravitational/teleport/integration.TestKube/TrustedClustersSNI
pass: github.com/gravitational/teleport/integration.TestKube/Disconnect/#00_iter_0
fail: github.com/gravitational/teleport/integration.TestKube/Disconnect/#01_iter_0
fail: github.com/gravitational/teleport/integration.TestKube/Disconnect
fail: github.com/gravitational/teleport/integration.TestKube
fail: github.com/gravitational/teleport/integration
===================================================
6 tests passed. 5 failed, 0 skipped
===================================================
FAIL: github.com/gravitational/teleport/integration
FAIL: github.com/gravitational/teleport/integration.TestKube
FAIL: github.com/gravitational/teleport/integration.TestKube/Deny
FAIL: github.com/gravitational/teleport/integration.TestKube/Disconnect/#01_iter_0
FAIL: github.com/gravitational/teleport/integration.TestKube/Disconnect
```

**Comment by russjones** [comment]
@NajiObeid Can you look into the issue @tcsc mentioned and try and repro?

**Comment by tcsc** [comment]
Running the 8.1.0 tests against the same cluster still fails - but differently:

- TestKube/Exec fails intermittently
- TestKube/Deny passes consistently
- TestKube/Disconnect/#00_iter_0 fails intermittently
- TestKube/Disconnect/#01_iter_0 fails intermittently

I suspect that the test differences are not a regression from this PR, but test rot because we haven't been running them as often as we should.

**Review by r0mant — COMMENTED** [review]
@NajiObeid Any luck looking into the issue mentioned by Trent above? Have you tested this against a GKE cluster? We'd like to merge this but make sure there are no regressions (difference in test behavior mentioned by Trent is a little worrying). Thanks.

**Comment by NajiObeid** [comment]
sorry I was off on monday and tuesday. Looking into it.

**Comment by NajiObeid** [comment]
@r0mant @tcsc 
`TestKube/Deny` failures seem to be caused by https://github.com/gravitational/teleport/pull/9759.
I have no other issues running the tests on my machine. I also have never ran the tests anywhere in the cloud.

**Comment by r0mant** [comment]
@smallinsky Could you take a look `TestKube/Deny` please and see if it needs updating due to https://github.com/gravitational/teleport/pull/9759 or if that's an issue with implementation (see comments above). Thanks!

**Review by russjones — APPROVED** [review]

**Comment by smallinsky** [comment]
Confirmed. After  https://github.com/gravitational/teleport/pull/9759 the Kube access evaluated based on v3 role roles  seems to be not backward compatible: I have created PR with a problem description:  https://github.com/gravitational/teleport/pull/10121

**Pull request merged**
